plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -802,73 +802,99 @@ func (checker *PodAffinityChecker) InterPodAffinityMatches(pod *api.Pod, meta in
 	if err != nil {
 		return false, err
 	}
-	if checker.NodeMatchPodAffinityAntiAffinity(pod, allPods, node) {
-		return true, nil
+	affinity, err := api.GetAffinityFromPodAnnotations(pod.Annotations)
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the current node match the inter-pod affinity scheduling constraints.
+	// Hard inter-pod affinity is not symmetric, check only when affinity.PodAffinity exists.
+	if affinity.PodAffinity != nil {
+		if !checker.NodeMatchesHardPodAffinity(pod, allPods, node, affinity.PodAffinity) {
+			return false, ErrPodAffinityNotMatch
+		}
+	}
+
+	// Hard inter-pod anti-affinity is symmetric, we should always check it.
+	if !checker.NodeMatchesHardPodAntiAffinity(pod, allPods, node, affinity.PodAntiAffinity) {
+		return false, ErrPodAffinityNotMatch
 	}
-	return false, ErrPodAffinityNotMatch
+
+	return true, nil
 }
 
 // AnyPodMatchesPodAffinityTerm checks if any of given pods can match the specific podAffinityTerm.
-func (checker *PodAffinityChecker) AnyPodMatchesPodAffinityTerm(pod *api.Pod, allPods []*api.Pod, node *api.Node, podAffinityTerm api.PodAffinityTerm) (bool, error) {
+// TODO: Do we really need any pod matching, or all pods matching? I think the latter.
```

Comment:
Why do you think it should be all pods matching? 

When called from NodeMatchesHardPodAffinity() it is enforcing a rule like "only put the new pod on a node that is already running a pod from the storage service, since the new pod needs very low-latency communication with the storage service." You don't need all the pods on the node to match, just one, for the node to be accepted.

When called from NodeMatchesHardPodAntiAffinity() it is enforcing a rule like like "do not put the new pod on a node that is already running (some pod that the new pod would conflict with)." You don't need all the pods on the node to match, just one, for the node to be rejected.